False positive for F841

Today the latest version of ruff (0.0.128) started  complaining about this test code:
```python
    def test_func_default_with_counter(self) -> None:
        counter = 0
        def next_value() -> int:
            nonlocal counter
            counter += 1
            return counter
        class HasFuncDefaultInt(Model):
            value = Int(default=next_value)
        obj1 = HasFuncDefaultInt()
        obj2 = HasFuncDefaultInt()
        assert obj1.value + 1 == obj2.value
```
with 
> tests/unit/bokeh/test_objects.py:291:9: F841 Local variable `counter` is assigned to but never used

It seems  the lasts`ruff` does not understand the usage with `nonlocal`. Previous versions of `ruff` (and `flake8`) do not warn on this code. 

(this isn't code we'd ever use in the library itself, but we do have to handle whatever crazy cases users might throw at us, and this is a test for one of those)

--- a/resources/test/fixtures/F841.py
+++ b/resources/test/fixtures/F841.py
@@ -35,3 +35,20 @@ def f4():
     _ = 1
     __ = 1
     _discarded = 1
+
+
+a = 1
+
+
+def f5():
+    global a
+
+    # Used in `f7` via `nonlocal`.
+    b = 1
+
+    def f6():
+        # F841
+        b = 1
+
+    def f7():
+        nonlocal b

--- a/src/ast/types.rs
+++ b/src/ast/types.rs
@@ -83,6 +83,7 @@ pub enum BindingKind {
     Binding,
     LoopVar,
     Global,
+    Nonlocal,
     Builtin,
     ClassDefinition,
     Definition,

--- a/src/check_ast.rs
+++ b/src/check_ast.rs
@@ -213,20 +213,73 @@ where
 
         // Pre-visit.
         match &stmt.node {
-            StmtKind::Global { names } | StmtKind::Nonlocal { names } => {
-                let global_scope_id = self.scopes[GLOBAL_SCOPE_INDEX].id;
-                let scope =
-                    &mut self.scopes[*(self.scope_stack.last().expect("No current scope found."))];
-                if scope.id != global_scope_id {
+            StmtKind::Global { names } => {
+                let scope_index = *self.scope_stack.last().expect("No current scope found.");
+                if scope_index != GLOBAL_SCOPE_INDEX {
+                    let scope = &mut self.scopes[scope_index];
+                    let usage = Some((scope.id, Range::from_located(stmt)));
                     for name in names {
+                        // Add a binding to the current scope.
                         scope.values.insert(
                             name,
                             Binding {
                                 kind: BindingKind::Global,
-                                used: Some((global_scope_id, Range::from_located(stmt))),
+                                used: usage,
                                 range: Range::from_located(stmt),
                             },
                         );
+                    }
+
+                    // Mark the binding in the global scope as used.
+                    for name in names {
+                        if let Some(mut existing) = self.scopes[GLOBAL_SCOPE_INDEX]
+                            .values
+                            .get_mut(&name.as_str())
+                        {
+                            existing.used = usage;
+                        }
+                    }
+                }
+
+                if self.settings.enabled.contains(&CheckCode::E741) {
+                    let location = Range::from_located(stmt);
+                    self.add_checks(
+                        names
+                            .iter()
+                            .filter_map(|name| {
+                                pycodestyle::checks::ambiguous_variable_name(name, location)
+                            })
+                            .into_iter(),
+                    );
+                }
+            }
+            StmtKind::Nonlocal { names } => {
+                let scope_index = *self.scope_stack.last().expect("No current scope found.");
+                if scope_index != GLOBAL_SCOPE_INDEX {
+                    let scope = &mut self.scopes[scope_index];
+                    let usage = Some((scope.id, Range::from_located(stmt)));
+                    for name in names {
+                        // Add a binding to the current scope.
+                        scope.values.insert(
+                            name,
+                            Binding {
+                                kind: BindingKind::Global,
+                                used: usage,
+                                range: Range::from_located(stmt),
+                            },
+                        );
+                    }
+
+                    // Mark the binding in the defining scopes as used too. (Skip the global scope
+                    // and the current scope.)
+                    for name in names {
+                        for index in self.scope_stack.iter().skip(1).rev().skip(1) {
+                            if let Some(mut existing) =
+                                self.scopes[*index].values.get_mut(&name.as_str())
+                            {
+                                existing.used = usage;
+                            }
+                        }
                     }
                 }
 

